app/daemon.go

```diff
@@ -206,15 +233,17 @@ func (d *MenderShellDaemon) messageLoop() (err error) {
 				"messageLoop: error on readMessage: %v; disconnecting, waiting for reconnect.",
 				err,
 			)
-			connectionmanager.Close(ws.ProtoTypeShell)
-			e := MenderShellDaemonEvent{
-				event: EventReconnectRequest,
+			// nolint:lll
+			// If we used a closed connection means that it has been closed from other goroutine
+			// and a reconnection is ongoing (or done). Just wait for the event and continue
+			// This can happens whe dbusEventLoop detects a change in ServerURL and/or JWT token.
+			// It should be safe to use this string, see:
+			// https://github.com/golang/go/blob/529939072eef730c82333344f321972874758be8/src/net/error_test.go#L502-L507 or about
+			if !strings.Contains(err.Error(), "use of closed network connection") {
+				connectionmanager.Close(ws.ProtoTypeShell)
+				sendConnectReq = true
```

Comment:
I have some doubts here. this was on purpose, although I am not saying it was necessarily right :), do you really want _not_ to close the connection and reconnect on read error?

One possible error scenario returned from Read is: NextReader returning error, gorilla says:
```go
// [...] Errors returned from this method are
// permanent. Once this method returns a non-nil error, all subsequent calls to
// this method return the same error.
```

the other is the ReadAll on the reader. which is the network Read usually. do you think it is safer to close connection and reconnect on any error?

Reply:
second comment: 
we set _sendConnectReq = true_ only when there was not a certain error. but we set _waitConnectResp = true_ (lower) always. if waitConnectResp is true the messageLoop waits for message on channel `d.connectionEstChan`, this is only sent on `EventReconnect` in `eventLoop` on `eventChan` which only happens when `needsReconnect && d.authorized` in `dbusEventLoop`, the `d.authorized` is true always if we have a token, and `needsReconnect` is set to true on a message on `reconnectChan` in `dbusEventLoop` or if we detect a change in the token or server url, the message on `reconnectChan` is sent only from messageLoop if `sendConnectReq` is true.

Reply:
@oleorhagen while I greatly appreciate the work Lluis did, this code now becomes more and more hard to follow. it was ok at the beginning, but after the recent massive changes on the client side the situation is much different. 

it is really a simple problem:
1. we need one go routine which permanently gets the JWT and server URL over D-Bus. if any of those is empty or changes, it sends reconnect signal.
2. we need to wait for reconnect events, and start reconnecting, each time using `d.jwtToken` and `d.serverURL` (need to be protected by mutex) until we are connected
3. we need to close connection and start reconnecting on errors

what do you think, maybe a small rewrite?

Reply:
Be aware that we need to cherry-pick the bugfix to stable branches. If we do a rewrite, consider doing it on a separate commit and cherry-picking only the first one (the one I wrote + potential corrections) to the stable branches.

Reply:
@merlin-northern I don't have the mental capacity right now to follow your question. The point is that there are two ways of closing the connection, from `dbusEventLoop` on changes to url or token, or here on remote errors. This `if` is for the former case (we will be getting "closed network connection" error) in which case the connect request has already been sent and we only need to wait for the reconnect.

Reply:
> it is really a simple problem:
> 
>     1. we need one go routine which permanently gets the JWT and server URL over D-Bus. if any of those is empty or changes, it sends reconnect signal.
> 
>     2. we need to wait for reconnect events, and start reconnecting, each time using `d.jwtToken` and `d.serverURL` (need to be protected by mutex) until we are connected
> 
>     3. we need to close connection and start reconnecting on errors
> 

The issue is that 1 and 3 will happen at the same time: On changes to the auth parameters, the mender client will close the websocket connection to mender-connect and send a D-Bus signal with the new parameters. That is exactly why we have the catch for "closed network connection" error, as this indicates that the D-Bus loop has gotten the event before the message loop.